go-controller/pkg/util/pod_annotation.go

```diff
@@ -301,12 +302,30 @@ func GetPodCIDRsWithFullMask(pod *v1.Pod, nInfo NetInfo) ([]*net.IPNet, error) {
 }
 
 // GetPodIPsOfNetwork returns the pod's IP addresses, first from the OVN annotation
-// and then falling back to the Pod Status IPs. This function is intended to
-// also return IPs for HostNetwork and other non-OVN-IPAM-ed pods.
+// and then falling back to IPs from "k8s.ovn.org/pod-networks" pod annotation if pod
+// is served by user defined primary network, otherwise retrieve it from Pod Status IPs.
+// This function is intended to also return IPs for HostNetwork and other non-OVN-IPAM-ed pods.
 func GetPodIPsOfNetwork(pod *v1.Pod, nInfo NetInfo) ([]net.IP, error) {
 	if nInfo.IsSecondary() {
 		return SecondaryNetworkPodIPs(pod, nInfo)
 	}
+	if nInfo.IsPrimaryNetwork() && !PodWantsHostNetwork(pod) {
+		nadList := nInfo.GetNADs()
+		for _, nad := range nadList {
+			userNet, err := UnmarshalPodAnnotation(pod.Annotations, nad)
+			if err != nil && strings.Contains(err.Error(), "no ovn pod annotation for network") {
+				// NAD belongs to same network, but serving different namespace. so skip processing it.
+				continue
+			} else if err != nil {
+				return nil, err
+			}
+			ips := make([]net.IP, 0, len(userNet.IPs))
+			for _, netIP := range userNet.IPs {
+				ips = append(ips, netIP.IP)
+			}
+			return ips, nil
+		}
+	}
```

Comment:
Please note that `nInfo.IsSecondary` return `true` for primary networks as well.

However, how is this implementation any different functionally than `SecondaryNetworkPodIPs`?

Reply:
> Please note that `nInfo.IsSecondary` return `true` for primary networks as well.

yes @jcaamano , thanks for the catch, it was actually causing failures for upstream netpol tests. fixed it now.

> 
> However, how is this implementation any different functionally than `SecondaryNetworkPodIPs`?

The retrieval of IPs from `SecondaryNetworkPodIPs` is retrieving nad name based on `k8s.v1.cni.cncf.io/networks` annotation which doesn't work for primary networks. hence separate handling is needed for primary networks.

Reply:
Can you add a comment?

Reply:
done.